README.md

# @rowanmanning/app

Build Express applications with all of my preferences baked in.

## Should I use this?

Not really. Unless you're Rowan Manning. Even then it's not advisable because this does too much - use [express-config](https://github.com/rowanmanning/express-config) instead.


## Table of Contents

  * [Requirements](#requirements)
  * [Usage](#usage)
    * [Creating an app](#creating-an-app)
    * [Examples](#examples)
  * [Contributing](#contributing)
  * [License](#license)


## Requirements

This library requires the following to run:

  * [Node.js](https://nodejs.org/) 14+


## Usage

Install with [npm](https://www.npmjs.com/):

```sh
npm install @rowanmanning/app
```

Load the library into your code with a `require` call:

```js
const App = require('@rowanmanning/app');
```

### Creating an app

To create an app, create an instance of the `App` class. `options` is an object, [available options are documented here](#app-options).

```js
const app = new App(options);
```

You may also want to [extend the `App` class](#extending) in order to add additional features or default some of the options to something that makes sense for you.

You'll also need to create the directory structure that your application needs. These folders are configurable, but the default directory structure is:

```
app
├── client
|   ├── public
|   └── sass
└── server
    ├── controller
    ├── model
    └── view
```

### Starting the app

To start the application, you'll need to use the `setup` method. This does not return anything, if you want to know when the application has started you can use [events](#app-events).

```js
app.setup();
```

### Stopping the app

You can stop the application safely with:

```js
app.teardown();
```

### App options

When you initialise a new application, the following options are available:

  - **`basePath`**: `String`. The path to look for application files in, prepended to all other paths. Defaults to `<CWD>`
  - **`controllerSubPath`**: `String`. The path to look for application controllers in. Will be prepended with `options.basePath`. Defaults to `server/controller`
  - **`databaseUrl`**: `String`. A MongoDB connection string to use for persistent storage. Defaults to `null`
  - **`enforceHttps`**: `Boolean`. Whether to use redirect HTTP requests to HTTPS. Defaults to `true` when `options.env` is `production`, and `false` when `options.env` is `development`
  - **`env`**: `String`. The environment that the application is running in. Defaults to the `NODE_ENV` environment variable, or `development` if it's not set
  - **`logger`**: `Object`. An object with log methods. Defaults to `console`
  - **`logger.info`**: `Function`. A function used to log information. Defaults to `console.info`
  - **`logger.error`**: `Function`. A function used to log errors. Defaults to `console.error`
  - **`logger.debug`**: `Function`. A function used to log debug messages. Defaults to `console.debug`
  - **`modelSubPath`**: `String`. The path to look for application models in. Will be prepended with `options.basePath`. Defaults to `server/model`
  - **`name`**: `String`. The name of the application. Used in logs. Defaults to `App`
  - **`port`**: `Number`. The port that the application should run on. Defaults to the `PORT` environment variable or `8080` if it's not set.
  - **`publicCacheMaxAge`**: `Number`. The cache max-age for public resources in production. Defaults to `604800000` (one week)
  - **`publicSubPath`**: `String`. The path to look for application public files in. Will be prepended with `options.basePath`. Defaults to `client/public`
  - **`requestLogFormat`**: `String`. The request log format, see the [Morgan documentation](https://github.com/expressjs/morgan) for more information. Defaults to `combined` when `options.env` is `production`, and `dev` when `options.env` is `development`
  - **`requestLogOutputStream`**: `Stream`. The stream to pipe request logs into. Defaults to `process.stdout`
  - **`securityConfig`**: `Object`. Security configuration to pass into [Helmet](https://helmetjs.github.io/). Defaults to `undefined`.
  - **`sassBundles`**: `Object`. A map of CSS URLs and Sass source paths, where each key is the URL path that the CSS bundle is served on, and each value is the location of the entry point Sass source file for that bundle, see [Resave Sass](https://github.com/rowanmanning/resave-sass) for more information. The source paths are relative to `options.basePath` and `options.sassSubPath`. Defaults to `{'/main.css': 'main.scss'}`
  - **`sassSubPath`**: The path to look for Sass files in. Will be prepended with `options.basePath`. Defaults to `client/sass`
  - **`sessionSecret`**: `String`. A secret used to sign session cookies with. If not set, sessions and flash messages are disabled. See [express-session](https://github.com/expressjs/session) and [connect-flash](https://github.com/jaredhanson/connect-flash) for documentation on these. Defaults to `null`
  - **`trustProxy`**: Express [trust proxy](http://expressjs.com/en/api.html#trust.proxy.options.table) settings. This is only used if `options.env` is "production". Defaults to `true`
  - **`useSecureCookies`**: `Boolean`. Whether to use secure cookies. Defaults to `true` when `options.env` is `production`, and `false` when `options.env` is `development`
  - **`viewSubPath`**: `String`. The path to look for application views in. Will be prepended with `options.basePath`. Defaults to `server/view`

### View data

The following variables are automatically added to `response.locals`, and are available in all view files:

  - **`app`**: `Object`. The `App` instance.
  - **`request`**: `Object`. The Express `request` object that resulted in the render.
  - **`currentUrl`**: `String`. The Express `request.url`.
  - **`currentPath`**: `String`. The Express `request.path`.

### App events

The `App` class extends [`EventEmitter`](https://nodejs.org/api/events.html), and you can listen on the following events:

  - **`database:connected`**: emitted when the MongoDB database connection has been set up.
  - **`server:created`**: emitted when the node HTTP server has been set up. Called with the created HTTP server
  - **`server:started`**: emitted when the HTTP server has started listening on a port. Called with the HTTP server
  - **`setup:error`**: emitted when the `setup` method fails to set up the application. In this scenario it is no longer possible to start the application so it should be [torn down](#stopping-the-app)

### Extending

It's possible to extend the `App` class to add additional features or provide default options:

```js
class MyApp extends App {
    constructor(options = {}) {
        options.name = 'My App';
        super(options);
    }
}
```

The following methods can also be overridden to add or change behaviour. They are called in this order during startup:

  - **`setupClientAssetCompilation`**: Set up client-side asset compilation middleware
  - **`setupControllers`**: Load controllers from the file system
  - **`setupDatabase`**: Initialise the MongoDB database. See also the `database:connected` event
  - **`setupExpress`**: Initialise express and an HTTP server
  - **`setupModels`**: Load models from the file system and add them to `app.models`
  - **`startServer`**: Start the HTTP server


### Examples

A ready-to-use main application file looks something like this:

```js
const App = require('@rowanmanning/app');

const app = new App({
    basePath: __dirname,
    databaseUrl: process.env.DATABASE_URL,
    name: 'My Application',
    sessionSecret: process.env.SESSION_SECRET
});

// Catch setup errors
app.once('setup:error', error => {
    process.exitCode = 1;
    app.log.error(error.stack);
    app.teardown();
});

// Set up the application
app.setup();
```

There are also a few example implementations in the repo which demonstrate some features:

  - **[Basic](example/basic)**: the simplest use of the library to render pages. Run `npm run example-basic` to start the application and visit [localhost:8080](http://localhost:8080/).

  - **[TODO](example/todo)**: a TODO-list application. Run `npm run example-todo` to start the application and visit [localhost:8080](http://localhost:8080/).


## Contributing

[The contributing guide is available here](docs/contributing.md). All contributors must follow [this library's code of conduct](docs/code_of_conduct.md).


## License

Licensed under the [MIT](LICENSE) license.<br/>
Copyright &copy; 2019, Rowan Manning


CHANGELOG.md
# Changelog

## [1.0.0-alpha.33](https://github.com/rowanmanning/app/compare/v1.0.0-alpha.32...v1.0.0-alpha.33) (2022-08-26)


### Bug Fixes

* bump mongoose from 6.5.0 to 6.5.2 ([9bec598](https://github.com/rowanmanning/app/commit/9bec59847ec19465a26efd11ba9fb47ca5bd2f03))
* bump mongoose from 6.5.2 to 6.5.3 ([#328](https://github.com/rowanmanning/app/issues/328)) ([2d8a09e](https://github.com/rowanmanning/app/commit/2d8a09e7265a67438b3eb0db12233bdbb4985e4b))
* bump preact from 10.10.0 to 10.10.2 ([02bf950](https://github.com/rowanmanning/app/commit/02bf950dba5d95c684cc2b79d1f2a60361d549e6))
* bump preact from 10.10.2 to 10.10.3 ([3df0376](https://github.com/rowanmanning/app/commit/3df0376d9906ec3de0e63b75cced50fbd38a98dd))
* bump preact from 10.10.3 to 10.10.4 ([0365e62](https://github.com/rowanmanning/app/commit/0365e627d5b7f5d9b5f699fcf81e4d8d9549b970))
* bump preact from 10.10.4 to 10.10.6 ([19eaec3](https://github.com/rowanmanning/app/commit/19eaec37f6965c1c7af9cbd0ceeea56dcd6aacd6))

## [1.0.0-alpha.32](https://github.com/rowanmanning/app/compare/v1.0.0-alpha.31...v1.0.0-alpha.32) (2022-08-02)


### Bug Fixes

* bump helmet from 5.1.0 to 5.1.1 ([5361314](https://github.com/rowanmanning/app/commit/5361314d33526a3c98056311d0f4294058c842e6))
* bump mongoose from 6.4.6 to 6.5.0 ([b0823e5](https://github.com/rowanmanning/app/commit/b0823e53c8bf928fa0f2de99637e05118bdc87b3))

## [1.0.0-alpha.31](https://github.com/rowanmanning/app/compare/v1.0.0-alpha.30...v1.0.0-alpha.31) (2022-07-21)


### Bug Fixes

* bump mongoose from 6.4.4 to 6.4.5 ([15f5199](https://github.com/rowanmanning/app/commit/15f5199f902c738e46e05455e38e95d9da3f6c99))
* bump mongoose from 6.4.5 to 6.4.6 ([8ea053b](https://github.com/rowanmanning/app/commit/8ea053bffa739d583a44ee01fe0ee0dddfb60b98))
* bump preact from 10.9.0 to 10.10.0 ([5d8db4f](https://github.com/rowanmanning/app/commit/5d8db4f3d8b2eb43feaf37087d1fca5267a9a918))
* hide the misc changelog section ([f7be95e](https://github.com/rowanmanning/app/commit/f7be95ef6a918da50ec01655947b87b05aded3a5))

## [1.0.0-alpha.30](https://github.com/rowanmanning/app/compare/v1.0.0-alpha.29...v1.0.0-alpha.30) (2022-07-09)


### Features

* bump mongoose from 6.3.8 to 6.4.2 ([b9c96f4](https://github.com/rowanmanning/app/commit/b9c96f4d2fbe01850ccb8660ba92a04322e0530b))
* bump preact from 10.7.3 to 10.8.2 ([2af4c01](https://github.com/rowanmanning/app/commit/2af4c01ce85217f3ff21e35558bde24a1232322b))
* bump preact from 10.8.2 to 10.9.0 ([7748e78](https://github.com/rowanmanning/app/commit/7748e788815f9192fb52e09b6694f56999a2fb22))


### Bug Fixes

* bump mongoose from 6.3.5 to 6.3.6 ([ca67601](https://github.com/rowanmanning/app/commit/ca67601a10cc7cab076bfdc14c59b3991e2ccd3d))
* bump mongoose from 6.3.6 to 6.3.8 ([d86a951](https://github.com/rowanmanning/app/commit/d86a9512f3afd8d2a265a4120ccaaf0e399e2528))
* bump mongoose from 6.4.2 to 6.4.4 ([63ed5c2](https://github.com/rowanmanning/app/commit/63ed5c2018294a29bffccba38f63cbc6603f4c35))
* bump preact from 10.7.2 to 10.7.3 ([e468db0](https://github.com/rowanmanning/app/commit/e468db0b016b9e18960b2ff8c681ba7c4e80807e))
* bump varname from 4.0.1 to 4.1.0 ([f064a78](https://github.com/rowanmanning/app/commit/f064a783d0431b19487fe38ce65d1445faa94080))


### Miscellaneous

* bump @commitlint/cli from 17.0.1 to 17.0.2 ([e54d349](https://github.com/rowanmanning/app/commit/e54d349076c2bbfdc39374eb8eb9672e6026bc5a))
* bump @commitlint/cli from 17.0.2 to 17.0.3 ([2e1152c](https://github.com/rowanmanning/app/commit/2e1152c65729b98c15c3b4f15692e0ac2ce6678f))
* bump @commitlint/config-conventional from 17.0.0 to 17.0.2 ([db4d6b6](https://github.com/rowanmanning/app/commit/db4d6b68ee94a29945321087fb3844dc1b79333d))
* bump @commitlint/config-conventional from 17.0.2 to 17.0.3 ([8409d8d](https://github.com/rowanmanning/app/commit/8409d8dcd737f503736fcddc66d820f9cdaf4eab))
* bump @rowanmanning/eslint-config from 3.5.0 to 4.0.2 ([756bf0c](https://github.com/rowanmanning/app/commit/756bf0c4f49eccbb6aedaada41f812a98664f96b))
* bump @rowanmanning/not-found from 3.0.3 to 3.1.0 ([617364e](https://github.com/rowanmanning/app/commit/617364eaf3463b4c5f5ea60f0c5fa5bca297273b))
* bump @rowanmanning/render-error-page from 3.0.1 to 3.1.0 ([374bbbe](https://github.com/rowanmanning/app/commit/374bbbea7730ee8463749837c34058416818e112))
* bump @rowanmanning/require-all from 2.1.1 to 2.2.0 ([9fba854](https://github.com/rowanmanning/app/commit/9fba8549509bffa9c2d4dd2ba939acd03fce7e32))
* bump eslint from 8.16.0 to 8.17.0 ([ec324a4](https://github.com/rowanmanning/app/commit/ec324a45f69fa63bb86066fd1838c14299072ea5))
* bump eslint from 8.17.0 to 8.19.0 ([ff7ff75](https://github.com/rowanmanning/app/commit/ff7ff75a762970f2e07e68d6b1b4c3fceffb12b7))
* bump nodemon from 2.0.16 to 2.0.18 ([8084cb7](https://github.com/rowanmanning/app/commit/8084cb7ab440f49fd1c47ad4a870c32a8d8e66b5))
* bump nodemon from 2.0.18 to 2.0.19 ([b69de3c](https://github.com/rowanmanning/app/commit/b69de3c6369f347bb43d0c051548eb5203e2d420))
* bump testdouble from 3.16.5 to 3.16.6 ([9f8b3b9](https://github.com/rowanmanning/app/commit/9f8b3b97e8fe6c0570219cad48f0e63899e09d16))
* fix eslint issues ([0dca7b2](https://github.com/rowanmanning/app/commit/0dca7b27dd845894b0f0d95bc7ecd39120ab42ba))

## [1.0.0-alpha.29](https://github.com/rowanmanning/app/compare/v1.0.0-alpha.28...v1.0.0-alpha.29) (2022-06-01)


### Features

* bump helmet from 5.0.2 to 5.1.0 ([f59dbe0](https://github.com/rowanmanning/app/commit/f59dbe09d686709c2e044d88aa22f7ea24c13668))


### Bug Fixes

* bump @rowanmanning/not-found from 2.1.0 to 3.0.2 ([53d3713](https://github.com/rowanmanning/app/commit/53d371338bb07b47b573d4c0a46ea6d8825b7db9))
* bump @rowanmanning/not-found from 3.0.2 to 3.0.3 ([3254864](https://github.com/rowanmanning/app/commit/3254864dca3721a60cca5e6a8ba91f44a0f86345))

lib/app.js
/**
 * @module @rowanmanning/app
 */
'use strict';

const {camelcase} = require('varname');
const {createServer} = require('http');
const EventEmitter = require('events');
const express = require('express');
const expressPreactViews = require('express-preact-views');
const flash = require('connect-flash');
const helmet = require('helmet');
const mongoose = require('mongoose');
const morgan = require('morgan');
const notFound = require('@rowanmanning/not-found');
const os = require('os');
const path = require('path');
const {redirectToHTTPS} = require('express-http-to-https');
const renderErrorPage = require('@rowanmanning/render-error-page');
const requireAll = require('@rowanmanning/require-all');
const resaveSass = require('resave-sass');
const session = require('express-session');
const MongoStore = require('connect-mongo');

/**
 * Represents an application.
 */
class App extends EventEmitter {

	/**
	 * Class constructor.
	 *
	 * @access public
	 * @param {object} [options={}]
	 *     An options object used to configure the application.
	 * @param {string} [options.basePath='<CWD>']
	 *     The path to look for application files in, prepended to all other paths.
	 * @param {string} [options.controllerSubPath='server/controller']
	 *     The path to look for application controllers in.
	 *     Will be prepended with `options.basePath`.
	 * @param {string} [options.databaseUrl=null]
	 *     A MongoDB connection string to use for persistent storage.
	 * @param {string} [options.enforceHttps]
	 *     Whether to use redirect HTTP requests to HTTPS. Defaults to `true` when `options.env` is
	 *     "production", and `false` when `options.env` is "development".
	 * @param {string} [options.env='development']
	 *     The environment that the application is running in. One of "production" or "development".
	 *     Defaults to the `NODE_ENV` environment variable.
	 * @param {object} [options.expressPreactOptions={beautify: true}]
	 *     Express Preact configuration options, see the documentation for more information:
	 *     {@see https://github.com/edwjusti/express-preact-views#options}.
	 * @param {object} [options.logger=console]
	 *     An object with log methods.
	 * @param {Function} [options.logger.info]
	 *     A function used to log information.
	 * @param {Function} [options.logger.error]
	 *     A function used to log errors.
	 * @param {Function} [options.logger.debug]
	 *     A function used to log debug notices.
	 * @param {string} [options.modelSubPath='server/model']
	 *     The path to look for application models in.
	 *     Will be prepended with `options.basePath`.
	 * @param {string} [options.name='App']
	 *     The name of the application. Used in logs.
	 * @param {(number | string)} [options.port=8080]
	 *     The port that the application should run on.
	 *     Defaults to the `PORT` environment variable.
	 * @param {number} [options.publicCacheMaxAge=604800000]
	 *     The cache max-age for public resources in production. Defaults to one week.
	 * @param {string} [options.publicSubPath='client/public']
	 *     The path to look for application public files in.
	 *     Will be prepended with `options.basePath`.
	 * @param {string} [options.requestLogFormat]
	 *     The request log format, see the Morgan documentation for more information:
	 *     {@see https://github.com/expressjs/morgan}.
	 *     Defaults to "combined" when `options.env` is "production",
	 *     and "dev" when `options.env` is "development".
	 * @param {import('stream').Stream} [options.requestLogOutputStream=process.stdout]
	 *     The stream to pipe request logs into.
	 * @param {object} [options.securityConfig]
	 *     Security configuration to pass into Helmet {@link https://helmetjs.github.io/}.
	 *     Defaults to undefined.
	 * @param {string} [options.sessionSecret=null]
	 *     A secret used to sign session cookies with. If not set, sessions are disabled.
	 * @param {(Array<string> | boolean | Function | number | string)} [options.trustProxy=true]
	 *     Express trust proxy settings, see the Express documentation for more information:
	 *     {@link http://expressjs.com/en/api.html#trust.proxy.options.table}.
	 *     This is only used if `options.env` is "production".
	 * @param {boolean} [options.useSecureCookies]
	 *     Whether to use secure cookies. Defaults to `true` when `options.env` is "production",
	 *     and `false` when `options.env` is "development".
	 * @param {string} [options.viewSubPath='server/view']
	 *     The path to look for application views in.
	 *     Will be prepended with `options.basePath`.
	 */
	constructor(options) {
		super();

		// Store key options as properties
		this.options = this.constructor.applyDefaultOptions(options);
		this.name = this.options.name;
		this.env = this.options.env;
		this.log = this.options.logger;

		// Resolve the application paths
		this.paths = {
			base: this.options.basePath,
			controller: path.resolve(this.options.basePath, this.options.controllerSubPath),
			model: path.resolve(this.options.basePath, this.options.modelSubPath),
			public: path.resolve(this.options.basePath, this.options.publicSubPath),
			sass: path.resolve(this.options.basePath, this.options.sassSubPath),
			view: path.resolve(this.options.basePath, this.options.viewSubPath)
		};

		// Set up places to store loaded in code
		this.controllers = {};
		this.models = {};
	}

	/**
	 * Set up the application.
	 *
	 * @access public
	 * @returns {App}
	 *     Returns the application instance.
	 */
	setup() {
		this.log.info(`[setup]: initialising in "${this.env}" mode`);
		try {
			this.setupDatabase();
			this.setupExpress();
			this.startServer();
			return this;
		} catch (error) {
			this.emit('setup:error', error);
			return this;
		}
	}

	/**
	 * Initialise the Mongo database. Used in class construction.
	 *
	 * @access private
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	setupDatabase() {
		if (!this.options.databaseUrl) {
			this.log.error('[setup:database]: missing "databaseUrl" option, database not set up');
			return;
		}
		this.db = mongoose.createConnection(this.options.databaseUrl);
		this.db.on('connected', () => {
			this.log.debug(`[setup:database]: connected`);
			this.emit('database:connected');
		});
		this.setupModels();
	}

	/**
	 * Initialise the application models. Used in class construction.
	 *
	 * @access private
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	setupModels() {
		try {
			for (const {name, moduleExports: setupModel} of requireAll(this.paths.model)) {
				const camelcaseName = camelcase(name);
				this.models[camelcaseName] = this.db.model(camelcaseName, setupModel(this));
				this.log.debug(`[setup:models]: "${camelcaseName}" model initialised`);
			}
		} catch (error) {
			error.message = `Models could not be loaded: ${error.message}`;
			throw error;
		}
	}

	/**
	 * Initialise the Express application. Used in class construction.
	 *
	 * @access private
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	setupExpress() {
		const app = this.express = this.router = express();
		const server = this.server = createServer(app);
		this.emit('server:created', server);

		// Configure Express
		app.enable('case sensitive routing');
		app.enable('strict routing');
		app.set('json spaces', 4);

		// Configure proxy trust in production
		if (this.env === 'production') {
			app.set('trust proxy', this.options.trustProxy);
		}

		// Render views with Preact
		app.set('views', this.paths.view);
		app.set('view engine', 'jsx');
		app.engine('jsx', expressPreactViews.createEngine(this.options.expressPreactOptions));

		// Provide some basic security settings with Helmet
		// and automatic HTTPS redirects
		app.use(helmet(this.options.securityConfig));
		if (this.options.enforceHttps) {
			app.use(redirectToHTTPS());
		}

		// Parse the request body
		app.use(express.urlencoded({
			extended: false
		}));
		app.use(express.json({
			strict: false
		}));

		// Set up sessions
		if (this.options.sessionSecret) {
			this.sessionMiddleware = session({
				cookie: {
					sameSite: 'lax',
					secure: this.options.useSecureCookies
				},
				name: `${this.name} Session`,
				resave: false,
				saveUninitialized: false,
				secret: this.options.sessionSecret,
				store: (
					this.options.databaseUrl ?
						MongoStore.create({client: this.db.getClient()}) :
						new session.Store()
				)
			});
			app.use(this.sessionMiddleware);
			this.log.info('[setup:sessions]: sessions set up successfully');

			// Set up flash messages
			this.flashMessageMiddleware = flash();
			app.use(this.flashMessageMiddleware);
		} else {
			this.log.error('[setup:sessions]: missing "sessionSecret" option, sessions not set up');
		}

		// Add some properties to the response locals
		app.locals.app = this;
		app.use((request, response, next) => {
			response.locals.request = request;
			response.locals.currentUrl = request.url;
			response.locals.currentPath = request.path;
			next();
		});

		// Add request logging
		if (this.options.requestLogFormat) {
			app.use(morgan(this.options.requestLogFormat, {
				stream: this.options.requestLogOutputStream,
				skip: request => request.path === '/favicon.ico'
			}));
		}

		// Load the controllers
		this.setupControllers();

		// Set up static routes
		const maxAge = (this.env === 'production' ? this.options.publicCacheMaxAge : 0);
		app.use(express.static(this.paths.public, {maxAge}));

		// Set up client-side asset compilation
		this.setupClientAssetCompilation();

		// Set up error handling
		app.use(notFound());
		app.use(renderErrorPage({
			errorLogger: error => {
				const stack = error.stack.split(/\n\s+/).slice(1).join('\n');
				this.log.error(
					`${error.name || 'Error'}: ${error.message} ${JSON.stringify({stack})}`
				);
			},
			errorLoggingFilter: error => {
				const status = error.statusCode || error.status;
				return !status || status >= 500;
			},
			includeErrorStack: (this.env !== 'production')
		}));
	}

	/**
	 * Initialise the application controllers. Used in class construction.
	 *
	 * @access private
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	setupControllers() {
		try {
			const controllers = requireAll(this.paths.controller);
			for (const {name, moduleExports: setupController} of controllers) {
				const camelcaseName = camelcase(name);
				this.controllers[camelcaseName] = setupController(this);
				this.log.debug(`[setup:controllers]: "${camelcaseName}" controller initialised`);
			}
		} catch (error) {
			error.message = `Controllers could not be loaded: ${error.message}`;
			throw error;
		}
	}

	/**
	 * Initialise the application client-side asset compiltion. Used in class construction.
	 *
	 * @access private
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	setupClientAssetCompilation() {

		// Set up Sass compilation
		this.router.use(resaveSass({
			basePath: this.paths.sass,
			bundles: this.options.sassBundles,
			log: {
				error: this.log.error.bind(this.log, '[assets:sass]:'),
				info: this.log.info.bind(this.log, '[assets:sass]:')
			},
			savePath: (this.env === 'production' ? this.paths.public : null)
		}));

	}

	/**
	 * Start the application server. Used in class construction.
	 *
	 * @access private
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	startServer() {
		this.server.listen(this.options.port, error => {
			if (error) {
				this.log.error(`[setup:server]: failed to start`);
				return this.emit('setup:error', error);
			}
			const address = `http://${os.hostname()}:${this.server.address().port}`;
			this.log.info(`[setup:server]: started successfully ${address}`);
			this.emit('server:started', this.server);
		});
	}

	/**
	 * Tear down the application.
	 *
	 * @access public
	 * @returns {undefined}
	 *     Returns nothing.
	 */
	teardown() {
		if (this.server) {
			this.server.close(() => {
				this.log.info('[teardown:server]: stopped successfully');
				delete this.server;
			});
		}
		if (this.db) {
			this.db.close(() => {
				this.log.info('[teardown:database]: closed connection successfully');
				delete this.db;
			});
		}
	}

	/**
	 * Apply default values to a set of user-provided options.
	 * Used internally by {@link App#constructor}.
	 *
	 * @access private
	 * @param {object} [userOptions={}]
	 *     Options to add on top of the defaults. See {@link App#constructor}.
	 * @returns {object}
	 *     Returns the defaulted options.
	 */
	static applyDefaultOptions(userOptions) {
		const options = Object.assign({}, App.defaultOptions, this.defaultOptions, userOptions);
		if (options.enforceHttps === undefined) {
			options.enforceHttps = (options.env === 'production');
		}
		if (options.requestLogFormat === undefined) {
			options.requestLogFormat = (options.env === 'production' ? 'combined' : 'dev');
		}
		if (options.useSecureCookies === undefined) {
			options.useSecureCookies = (options.env === 'production');
		}
		return options;
	}

}

/**
 * Default options to be used in construction of an application.
 *
 * @access private
 * @type {object}
 */
App.defaultOptions = {
	basePath: process.cwd(),
	controllerSubPath: 'server/controller',
	expressPreactOptions: {
		beautify: true
	},
	databaseUrl: null,
	enforceHttps: undefined,
	env: process.env.NODE_ENV || 'development',
	logger: console,
	modelSubPath: 'server/model',
	name: 'App',
	port: process.env.PORT || 8080,
	publicCacheMaxAge: 1000 * 60 * 60 * 24 * 7,
	publicSubPath: 'client/public',
	requestLogFormat: undefined,
	requestLogOutputStream: process.stdout,
	sassBundles: {
		'/main.css': 'main.scss'
	},
	sassSubPath: 'client/sass',
	sessionSecret: null,
	trustProxy: true,
	useSecureCookies: undefined,
	viewNamespacePaths: {},
	viewSubPath: 'server/view'
};

// Alias some dependencies so that they
// can be safely used by applications
App.Schema = mongoose.Schema;
App.SchemaType = mongoose.SchemaType;
App.ValidationError = mongoose.Error.ValidationError;

module.exports = App;
